pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -302,63 +303,68 @@ def normalize_url(url):
         raise exc
 
 
-def is_optimized_sync(repository, remote, url):
+def should_optimize_sync(sync_details, last_sync_details):
     """
     Check whether it is possible to optimize the synchronization or not.
 
-    Caution: we are not storing when the remote was last updated, so the order of this
-    logic must remain in this order where we first check the version number as other
-    changes than sync could have taken place such that the date or repo version will be
-    different from last sync.
-
     Args:
-        repository(RpmRepository): An RpmRepository to check optimization for.
-        remote(RpmRemote or UlnRemote): An RPMRemote or UlnRemote to check optimization for.
-        url(str): A remote repository URL.
+        sync_details (dict): A collection of details about the current sync configuration.
+        last_sync_details (dict): A collection of details about the previous sync configuration.
 
     Returns:
         bool: True, if sync is optimized; False, otherwise.
 
     """
-    with tempfile.TemporaryDirectory("."):
-        try:
-            result = get_repomd_file(remote, url)
-            repomd_path = result.path
-            repomd = cr.Repomd(repomd_path)
-            repomd_checksum = get_sha256(repomd_path)
-        except Exception:
-            return False
-
-    is_optimized = (
-        repository.last_sync_remote
-        and remote.pk == repository.last_sync_remote.pk
-        and repository.last_sync_repo_version == repository.latest_version().number
-        and remote.pulp_last_updated <= repository.latest_version().pulp_created
-        and is_previous_version(repomd.revision, repository.last_sync_revision_number)
-        and repository.last_sync_repomd_checksum == repomd_checksum
+    might_download_content = (
+        last_sync_details.get("download_policy") != "immediate"
+        and sync_details["download_policy"] == "immediate"
+    )
+    might_create_publication = (
```

Comment:
yeah, grab the latest version and create publication for it in case the policy has changed from X to mirror_complete

https://github.com/pulp/pulp_rpm/blob/26694ba56cb3ea23cc17ec97d96482331d538d5f/pulp_rpm/app/tasks/synchronizing.py#L536
```

if version or (last_sync_details.get("sync_policy") != SYNC_POLICIES.MIRROR_COMPLETE
        and sync_details["sync_policy"] == SYNC_POLICIES.MIRROR_COMPLETE
    ):
```

Reply:
looks good ^

Reply:
I just realized that

https://github.com/pulp/pulp_rpm/pull/2148/files#diff-fb406f31027edc1c5b65abe7635bbd2830a7cc05c24a487344d7649017d14628R539

In combination with

https://github.com/pulp/pulp_rpm/pull/2148/files#diff-fb406f31027edc1c5b65abe7635bbd2830a7cc05c24a487344d7649017d14628R510-R511

Is almost certainly broken.  If a sub-repo doesn't change and no new version is created then we won't be including it in the re-publish.